hibernate-vector/src/main/java/org/hibernate/vector/OracleByteVectorJdbcType.java

```diff
@@ -0,0 +1,182 @@
+package org.hibernate.vector;
+
+import java.sql.CallableStatement;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.BitSet;
+
+import org.hibernate.dialect.Dialect;
+import org.hibernate.sql.ast.spi.SqlAppender;
+import org.hibernate.type.SqlTypes;
+import org.hibernate.type.descriptor.ValueBinder;
+import org.hibernate.type.descriptor.ValueExtractor;
+import org.hibernate.type.descriptor.WrapperOptions;
+import org.hibernate.type.descriptor.java.JavaType;
+import org.hibernate.type.descriptor.jdbc.ArrayJdbcType;
+import org.hibernate.type.descriptor.jdbc.BasicBinder;
+import org.hibernate.type.descriptor.jdbc.BasicExtractor;
+import org.hibernate.type.descriptor.jdbc.JdbcLiteralFormatter;
+import org.hibernate.type.descriptor.jdbc.JdbcType;
+
+import oracle.jdbc.OracleConnection;
+import oracle.jdbc.OracleType;
+
+/**
+ * Specialized type mapping for single-byte integer vector {@link SqlTypes#VECTOR_INT8} SQL data type for Oracle.
+ *
+ * @author Hassan AL Meftah
+ */
+public class OracleByteVectorJdbcType extends ArrayJdbcType {
+
+
+	private static final byte[] EMPTY = new byte[0];
+
+	public OracleByteVectorJdbcType(JdbcType elementJdbcType) {
+		super( elementJdbcType );
+	}
+
+	@Override
+	public <T> JdbcLiteralFormatter<T> getJdbcLiteralFormatter(JavaType<T> javaTypeDescriptor) {
+		return null;
+	}
+
+	@Override
+	public void appendWriteExpression(String writeExpression, SqlAppender appender, Dialect dialect) {
+		appender.append( "vector(" );
+		appender.append( writeExpression );
+		appender.append( ", *, INT8)" );
+	}
+
+	@Override
+	public String getFriendlyName() {
+		return "VECTOR_INT8";
+	}
+
+	@Override
+	public int getJdbcTypeCode() {
+		return SqlTypes.VECTOR_INT8;
+	}
+
+	@Override
+	public String toString() {
+		return "OracleVectorTypeDescriptor";
+	}
+
+	private boolean isDriverBefore23(WrapperOptions options) throws SQLException {
+		final OracleConnection oracleConnection = options.getSession()
+				.getJdbcCoordinator()
+				.getLogicalConnection()
+				.getPhysicalConnection()
+				.unwrap( OracleConnection.class );
+
+		int majorVersion = oracleConnection.getMetaData().getDriverMajorVersion();
+		int minorVersion = oracleConnection.getMetaData().getDriverMinorVersion();
+
+		return ( majorVersion < 23 ) || ( majorVersion == 23 && minorVersion < 4 );
+	}
+
+	@Override
+	public <X> ValueBinder<X> getBinder(final JavaType<X> javaTypeDescriptor) {
+		return new BasicBinder<>( javaTypeDescriptor, this ) {
+			@Override
+			public void doBind(PreparedStatement st, X value, int index, WrapperOptions options) throws SQLException {
+				if ( isDriverBefore23( options ) ) {
+					st.setString( index, getStringVector( value, options ) );
+				}
+				else {
+					st.setObject( index, value, OracleType.valueOf( "VECTOR_INT8" ) );
```

Comment:
I'm wondering if we could replace the `valueOf` call with `OracleType.VECTOR_INT8`? Or does this code need to compile with older versions of Oracle JDBC? If so, then I get it: OracleType.VECTOR_INT8 doesn't exist in versions before 23.4.

Reply:
You are right, these types only register if OracleDialect is detected, and the user may have an older version of Oracle JDBC, that's why I used `valueOf` method instead of `OracleType.VECTOR_INT8`